---
title: "Loci discovery for target capture sequencing of ddRAD loci"
output: github_document
---

We first did ddRADseq on eight mitochondrially divergent Ophiontous individuals. Raw ddRAD reads were called and processed by AGRF using their genotyping-by-sequencing (GBS) in house Perl script based on Stacks v1.47. We start the filtering from the vcf file that AGRF returned to us (already filtered by read depth minimum 10x and variant quality scaore of 30). 
<br>
<br>
Keep SNPs only present in all eight individuals, as well as a minor allele frequency of at least 0.1, using vcftools
```{bash, error=TRUE, results='hide'}
module load vcftools/0.1.13

reads_dir=./ophbaits/snps/CAGRF17794.vcf
out_dir=./ophbaits/snps/filtering/vcf
log_file=./ophbaits/snps/filtering/vcf/vcf.oe

vcftools --vcf $reads_dir  --max-missing 1 --maf 0.01 --recode --recode-INFO-all --out $out_dir/filtered.vcf &> $log_file
```

Check the filtered vcf file to confirm missing data per loci and missing data per indv
```{bash, error=TRUE, results='hide'}
vcftools --vcf $out_dir/filtered.vcf.recode.vcf --missing-site

vcftools --vcf $out_dir/filtered.vcf.recode.vcf  --missing-indv
```

## Summary  
Filtered vcf file using vcftools shows 18237 SNPs & 8113 loci shared between all eight indv
<br>
<br>
Now to design baits based on consensus loci sequences between individuals (i.e. consensus loci sequences are located in the (unfiltered) loci.fasta file generated from --fasta-loci output in Stacks/population)
<br>
<br>
I used Geneious to filter out the loci to keep (those we identified from vcftools) in loci.fasta