SNP filtering - for dataset 3 (include all O. hexactis samples only)
================

Remove O. hexactis samples from the vcf file we already filtered for
indels:

``` bash
cd ./ophionotus/SNPfiltering/filtering3/1_rm_vic

vcftools --vcf ./ophionotus/SNPfiltering/filtering1/1_rm_indels/rm_indels.vcf.recode.vcf --keep hex.tsv --out hex_only --recode --recode-INFO-all
```

Remove samples with high missing data (>80%):

``` bash
## We don't have O. hexactis samples with high missing data. Skip.
```

Filter by minimum base quality score of 30:

``` bash
cd ./ophionotus/SNPfiltering/filtering3/3_minQ/

vcftools --vcf ./ophionotus/SNPfiltering/filtering3/1_rm_vic/hex_only.recode.vcf --minQ 30 --out minQ30 --recode --recode-INFO-all
```

Remove sites with mean depth less than 10x (‐‐min-meanDP 10) and greater
than average\*2 (–max-meanDP 32):

``` bash
cd ./ophionotus/SNPfiltering/filtering3/4_site_depth

vcftools --vcf ./ophionotus/SNPfiltering/filtering3/3_minQ/minQ30.recode.vcf --min-meanDP 10 --max-meanDP 32 --out site_depth --recode --recode-INFO-all
```

Keep biallelic sites only:

``` bash
cd ./ophionotus/SNPfiltering/filtering3/5_biallelic_sites

vcftools --vcf ./ophionotus/SNPfiltering/filtering3/4_site_depth/site_depth.recode.vcf --min-alleles 2 --max-alleles 2 --out biallelic_sites --recode --recode-INFO-all
```

Keep sites present in 70% of all samples (i.e. filter by 30% missing
data):

``` bash
cd ./ophionotus/SNPfiltering/filtering3/6_max_missing07

vcftools --vcf /ophionotus/SNPfiltering/filtering3/6_biallelic_sites/biallelic_sites.recode.vcf --max-missing 0.7 --out maxmissing07 --recode --recode-INFO-all
```

Keep sites with minor allele frequency at least 0.02:

``` bash
cd ./ophionotus/SNPfiltering/filtering3/7_maf002

vcftools --vcf ./ophionotus/SNPfiltering/filtering3/6_max_missing07/maxmissing07.recode.vcf --maf 0.02 --out maf002_accurate --recode --recode-INFO-all
```

Keep sites with a maximum observed heterozygosity value of 0.5: Typical
for RAD-loci data, for refs see
<https://onlinelibrary.wiley.com/doi/full/10.1111/j.1755-0998.2010.02967.x>
& <https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.13314>

``` bash
## first, define site ID
bcftools annotate --set-id '%CHROM\_%POS' maf002.recode.vcf > maf002_setID.vcf
```

``` r
## calculate observed heterozygosity value per site
library(adegenet)
library(pegas)
library(vcfR)

### get a genind from the semi-filtered vcf file from bcftools

vcf <- read.vcfR("./maf002_setID.vcf")
oph_genind <- vcfR2genind(vcf)
summary_genind = summary(oph_genind)

### get the Hobs per SNP 
summary_genind$Hobs 

write.csv(summary_genind$Hobs, file = "Hobs_per_SNP.csv")

### you can then sort the values in excel and save a list of the SNP (with site ID) with >0.5
```

``` bash
## Filter sites with maximum observed heterozygosity of 0.5
cd ./ophionotus/SNPfiltering/filtering3/8_max_obs_het05

vcftools --vcf ./ophionotus/SNPfiltering/filtering3/7_maf002/maf002_setID.vcf --exclude site_obs_het_over_05.tsv --out max_het05.vcf --recode --recode-INFO-all
```

Get 1 SNP per locus: In here, 140 is the sequence legnth in
unique_consensus_loci.fa

``` bash
cd ./ophionotus/SNPfiltering/filtering3/9_thin140

vcftools --vcf ./ophionotus/SNPfiltering/filtering3/8_max_obs_het05/max_het05.vcf.recode.vcf --thin 140 --out hex_thin140.vcf --recode --recode-INFO-all
```

### For stairwayplot

Keep SNPs present in 50% of all samples, keep all SNPs, filter by minor
allele count of 1 and max ob. het of 0.7

``` bash
cd ./ophionotus/SNPfiltering/filtering3/demographic/7_max_missing05

vcftools --vcf /ophionotus/SNPfiltering/filtering3/6_biallelic_sites/biallelic_sites.recode.vcf --max-missing 0.5 --out maxmissing05 --recode --recode-INFO-all
```

``` bash
cd ./ophionotus/SNPfiltering/filtering3/demographic/8_mac01

vcftools --vcf /ophionotus/SNPfiltering/filtering3/demographic/7_max_missing05/maxmissing05.recode.vcf --mac 1 --out mac1 --recode --recode-INFO-all
```

Keep sites with a maximum observed heterozygosity value of 0.7: Typical
for RAD-loci data, for refs see
<https://onlinelibrary.wiley.com/doi/full/10.1111/j.1755-0998.2010.02967.x>
& <https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.13314>

``` bash
## first, define site ID
bcftools annotate --set-id '%CHROM\_%POS' mac1.recode.vcf > mac1_setID.vcf
```

``` r
## calculate observed heterozygosity value per site
library(adegenet)
library(pegas)
library(vcfR)

### get a genind from the semi-filtered vcf file from bcftools

vcf <- read.vcfR("./mac1_setID.vcf")
oph_genind <- vcfR2genind(vcf)
summary_genind = summary(oph_genind)

### get the Hobs per SNP 
summary_genind$Hobs 

write.csv(summary_genind$Hobs, file = "Hobs_per_SNP.csv")

### you can then sort the values in excel and save a list of the SNP (with site ID) with >0.5
```

``` bash
## Filter sites with maximum observed heterozygosity of 0.7
cd ./ophionotus/SNPfiltering/filtering3/demographic/9_max_obs_het07

vcftools --vcf ./maf007_setID.vcf --exclude site_obs_het_over_07.tsv --out max_het07.vcf --recode --recode-INFO-all
```

Genotype likeligood estimations for SFS-based analyses (stairwayplot)

``` bash
##O. hexactis from Brnasfield Mouth
cd ./angsd_stairway/hex_BM

BAMlist=./angsd_stairway/hex_BM.filelist
refloci=./unique_consensus_samtools_fadix/BrittleStar-input-seq-final.fas

singularity run $SING/angsd-0.940.sif angsd -bam $BAMlist -ref $refloci \
        -uniqueOnly 1 -remove_bads 1 -skipTriallelic 1 -doMajorMinor 1 -trim 5 -C 50 -baq 1 \
        -minMapQ 30 -minQ 20 -minInd 10 -setMinDepthInd 3 -doCounts 1 -doHWE 1 -maxHetFreq 0.8 \
        -doSaf 1 -out hex -anc $refloci -GL 1 -P 1

#do not fold in AGNSD, fold using dadi instead! 
singularity run $SING/angsd-0.940.sif realSFS hex.saf.idx > hex_BM.sfs

#add header - 21 unfolded "BM" - to the first line of hex_BM.sfs. 21 = 2N+1
sed '1s/^/21 unfolded "BM" \n/' hex_BM.sfs > hex_BM_header.sfs

#fold SFS in dadi then export final folded SFS for staitway plot
ipython

import dadi
import pylab

fs = dadi.Spectrum.from_file("./hex_BM_header.sfs")
folded = fs.fold()
data = folded.project([14])
dadi.Plotting.plot_1d_fs(data, show=True)
data.to_file('dadi_hex_BM.sfs')


##O. hexactis from other Antarctic islands
Did this in cd ./angsd_sfs/hex (03.Target_capture_SNPfiltering_dataset1.md) see L200

#add header - 61 unfolded "other" - to the first line of hex_rmBM_unfolded.sfs. 61 = 2N+1
sed '1s/^/61 unfolded "other"\n/' hex_rmBM_unfolded.sfs > hex_rmBM_unfolded_header.sfs


#fold SFS in dadi then export final folded SFS for staitway plot
ipython

import dadi
import pylab

fs = dadi.Spectrum.from_file("./hex_rmBM_unfolded_header.sfs")
folded = fs.fold()
data = folded.project([40])
dadi.Plotting.plot_1d_fs(data, show=True)
data.to_file('dadi_hex_other.sfs')
```
