---
title: "SNP filtering - for dataset 2 (include all O. victoriae samples only) "
output: github_document
---

Remove O. hexactis samples from the vcf file we already filtered for indels:
```{bash, error=TRUE, results='hide'}
cd ./ophionotus/SNPfiltering/filtering2/1_rm_hex

vcftools --vcf ./ophionotus/SNPfiltering/filtering1/1_rm_indels/rm_indels.vcf.recode.vcf --keep vic.tsv --out vic_only --recode --recode-INFO-all
```

Remove samples with high missing data (>80%):
```{bash, error=TRUE, results='hide'}
cd ./ophionotus/SNPfiltering/filtering2/2_rm_highimiss/

vcftools --vcf ./ophionotus/SNPfiltering/filtering2/1_rm_hex/vic_only.recode.vcf --remove vic_highimiss.tsv --out rm_vic_highimiss --recode --recode-INFO-all
```

Filter by minimum base quality score of 30:
```{bash, error=TRUE, results='hide'}
cd ./ophionotus/SNPfiltering/filtering2/3_minQ/

vcftools --vcf ./ophionotus/SNPfiltering/filtering2/2_rm_highimiss/rm_vic_highimiss.recode.vcf --minQ 30 --out minQ30 --recode --recode-INFO-all
```

Remove sites with mean depth less than 10x  (‐‐min-meanDP 10) and greater than average*2 (--max-meanDP 32):
```{bash, error=TRUE, results='hide'}
cd ./ophionotus/SNPfiltering/filtering2/4_site_depth

vcftools --vcf ./ophionotus/SNPfiltering/filtering2/3_minQ/minQ30.recode.vcf --min-meanDP 10 --max-meanDP 32 --out site_depth --recode --recode-INFO-all
```

Keep biallelic sites only:
```{bash, error=TRUE, results='hide'}
cd ./ophionotus/SNPfiltering/filtering2/5_biallelic_sites

vcftools --vcf ./ophionotus/SNPfiltering/filtering2/4_site_depth/site_depth.recode.vcf --min-alleles 2 --max-alleles 2 --out biallelic_sites --recode --recode-INFO-all
```

Keep sites present in 70% of all samples (i.e. filter by 30% missing data):
```{bash, error=TRUE, results='hide'}
cd ./ophionotus/SNPfiltering/filtering2/6_max_missing07

vcftools --vcf /ophionotus/SNPfiltering/filtering2/6_biallelic_sites/biallelic_sites.recode.vcf --max-missing 0.7 --out maxmissing07 --recode --recode-INFO-all
```

Keep sites with minor allele frequency at least 0.02:
```{bash, error=TRUE, results='hide'}
cd ./ophionotus/SNPfiltering/filtering2/7_maf002

vcftools --vcf ./ophionotus/SNPfiltering/filtering2/6_max_missing07/maxmissing07.recode.vcf --maf 0.02 --out maf002_accurate --recode --recode-INFO-all
```

Keep sites with a maximum observed heterozygosity value of 0.5:
Typical for RAD-loci data, for refs see https://onlinelibrary.wiley.com/doi/full/10.1111/j.1755-0998.2010.02967.x &
https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.13314
```{bash, error=TRUE, results='hide'}
## first, define site ID
bcftools annotate --set-id '%CHROM\_%POS' maf002.recode.vcf > maf002_setID.vcf
```
```{r, eval=F, echo=T}
## calculate observed heterozygosity value per site
library(adegenet)
library(pegas)
library(vcfR)

### get a genind from the semi-filtered vcf file from bcftools

vcf <- read.vcfR("./maf002_setID.vcf")
oph_genind <- vcfR2genind(vcf)
summary_genind = summary(oph_genind)

### get the Hobs per SNP 
summary_genind$Hobs 

write.csv(summary_genind$Hobs, file = "Hobs_per_SNP.csv")

### you can then sort the values in excel and save a list of the SNP (with site ID) with >0.5
```

```{bash, error=TRUE, results='hide'}
## Filter sites with maximum observed heterozygosity of 0.5
cd ./ophionotus/SNPfiltering/filtering2/8_max_obs_het05

vcftools --vcf ./ophionotus/SNPfiltering/filtering2/7_maf002/maf002_setID.vcf --exclude site_obs_het_over_05.tsv --out max_het05.vcf --recode --recode-INFO-all
```

Get 1 SNP per locus:
In here, 140 is the sequence legnth in unique_consensus_loci.fa
```{bash, error=TRUE, results='hide'}
cd ./ophionotus/SNPfiltering/filtering2/9_thin140

vcftools --vcf ./ophionotus/SNPfiltering/filtering2/8_max_obs_het05/max_het05.vcf.recode.vcf --thin 140 --out vic_thin140.vcf --recode --recode-INFO-all
```

### For stairwayplot
Keep SNPs present in 50% of all samples, keep all SNPs, filter by minor allele count of 1 and max ob. het of 0.7
```{bash, error=TRUE, results='hide'}
cd ./ophionotus/SNPfiltering/filtering2/demographic/7_max_missing05

vcftools --vcf /ophionotus/SNPfiltering/filtering2/6_biallelic_sites/biallelic_sites.recode.vcf --max-missing 0.5 --out maxmissing05 --recode --recode-INFO-all
```

```{bash, error=TRUE, results='hide'}
cd ./ophionotus/SNPfiltering/filtering2/demographic/8_mac01

vcftools --vcf /ophionotus/SNPfiltering/filtering2/demographic/7_max_missing05/maxmissing05.recode.vcf --mac 1 --out mac1 --recode --recode-INFO-all
```

Keep sites with a maximum observed heterozygosity value of 0.7:
Typical for RAD-loci data, for refs see https://onlinelibrary.wiley.com/doi/full/10.1111/j.1755-0998.2010.02967.x &
https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.13314
```{bash, error=TRUE, results='hide'}
## first, define site ID
bcftools annotate --set-id '%CHROM\_%POS' mac1.recode.vcf > mac1_setID.vcf
```
```{r, eval=F, echo=T}
## calculate observed heterozygosity value per site
library(adegenet)
library(pegas)
library(vcfR)

### get a genind from the semi-filtered vcf file from bcftools

vcf <- read.vcfR("./mac1_setID.vcf")
oph_genind <- vcfR2genind(vcf)
summary_genind = summary(oph_genind)

### get the Hobs per SNP 
summary_genind$Hobs 

write.csv(summary_genind$Hobs, file = "Hobs_per_SNP.csv")

### you can then sort the values in excel and save a list of the SNP (with site ID) with >0.5
```

```{bash, error=TRUE, results='hide'}
## Filter sites with maximum observed heterozygosity of 0.7
cd ./ophionotus/SNPfiltering/filtering2/demographic/9_max_obs_het07

vcftools --vcf ./maf007_setID.vcf --exclude site_obs_het_over_07.tsv --out vic_allsnps.vcf --recode --recode-INFO-all
```

Genotype likeligood estimations for SFS-based analyses (stairwayplot)
```{bash, error=TRUE, results='hide'}
##O. victoriae from deep continental shelf (>1000m)
cd ./angsd_stairway/vic_deep

BAMlist=./angsd_stairway/vic_deep.filelist
refloci=./unique_consensus_samtools_fadix/BrittleStar-input-seq-final.fas

singularity run $SING/angsd-0.940.sif angsd -bam $BAMlist -ref $refloci \
        -uniqueOnly 1 -remove_bads 1 -skipTriallelic 1 -doMajorMinor 1 -trim 5 -C 50 -baq 1 \
        -minMapQ 30 -minQ 20 -minInd 12 -setMinDepthInd 3 -doCounts 1 -doHWE 1 -maxHetFreq 0.8 \
	    -doSaf 1 -out vic -anc $refloci -GL 1 -P 10

#do not fold in AGNSD, fold using dadi instead! 
singularity run $SING/angsd-0.940.sif realSFS vic.saf.idx > vic_deep.sfs

#add header - 33 unfolded "deep" - to the first line of vic_deep.sfs. 33 = 2N+1
sed '1s/^/33 unfolded "deep"\n/' vic_deep.sfs > vic_deep_header.sfs

#fold SFS in dadi then export final folded SFS for staitway plot
ipython

import dadi
import pylab

fs = dadi.Spectrum.from_file("./vic_deep_header.sfs")
folded = fs.fold()
data = folded.project([18])
dadi.Plotting.plot_1d_fs(data, show=True)
data.to_file('./dadi_vic_deep.sfs')


##O. victoriae from Antarctic islands
cd ./angsd_stairway/vic_island

BAMlist=./angsd_stairway/vic_island.filelist
refloci=./unique_consensus_samtools_fadix/BrittleStar-input-seq-final.fas

singularity run $SING/angsd-0.940.sif angsd -bam $BAMlist -ref $refloci \
        -uniqueOnly 1 -remove_bads 1 -skipTriallelic 1 -doMajorMinor 1 -trim 5 -C 50 -baq 1 \
        -minMapQ 30 -minQ 20 -minInd 56 -setMinDepthInd 3 -doCounts 1 -doHWE 1 -maxHetFreq 0.8 \
	    -doSaf 1 -out vic -anc $refloci -GL 1 -P 10

#do not fold in AGNSD, fold using dadi instead! 
singularity run $SING/angsd-0.940.sif realSFS vic.saf.idx > vic_island.sfs

#add header - 141 unfolded "island" - to the first line of vic_island.sfs. 141 = 2N+1
sed '1s/^/141 unfolded "island"\n/' vic_island.sfs > vic_island_header.sfs

#fold SFS in dadi then export final folded SFS for staitway plot
ipython

import dadi
import pylab

fs = dadi.Spectrum.from_file("./vic_island_header.sfs")
folded = fs.fold()
data = folded.project([80])
dadi.Plotting.plot_1d_fs(data, show=True)
data.to_file('./dadi_vic_island.sfs')


##O. victoriae from shallow continental shelf (<1000m)
cd ./angsd_stairway/vic_shelf

BAMlist=./angsd_stairway/vic_shelf.filelist
refloci=./unique_consensus_samtools_fadix/BrittleStar-input-seq-final.fas

singularity run $SING/angsd-0.940.sif angsd -bam $BAMlist -ref $refloci \
        -uniqueOnly 1 -remove_bads 1 -skipTriallelic 1 -doMajorMinor 1 -trim 5 -C 50 -baq 1 \
        -minMapQ 30 -minQ 20 -minInd 57 -setMinDepthInd 3 -doCounts 1 -doHWE 1 -maxHetFreq 0.8 \
	    -doSaf 1 -out vic -anc $refloci -GL 1 -P 10

#do not fold in AGNSD, fold using dadi instead! 
singularity run $SING/angsd-0.940.sif realSFS vic.saf.idx > vic_shelf.sfs

#add header - 145 unfolded "shelf" - to the first line of vic_shelf.sfs. 145 = 2N+1
sed '1s/^/145 unfolded "shelf"\n/' vic_shelf.sfs > vic_shelf_header.sfs

#fold SFS in dadi then export final folded SFS for staitway plot
ipython

import dadi
import pylab

fs = dadi.Spectrum.from_file("./vic_shelf_header.sfs")
folded = fs.fold()
data = folded.project([92])
dadi.Plotting.plot_1d_fs(data, show=True)
data.to_file('./dadi_vic_shelf.sfs')
```