---
title: "Analyse population structure"
output: github_document
---

## Population genetic statistics (PopGenome)

Calculate Tajima’s D and Fu’s FS between localities of O. victoriae 

```{r, eval=F, echo=T}
library(PopGenome)
GENOME.class <- readData("./ophionotus/analysis/genetic_statistics/popgenome/vic_thin140.vcf", format="VCF", include.unknown=TRUE)
get.sum.data(GENOME.class)
GENOME.class@populations

get.individuals(GENOME.class)

AL <- c("IE_2009_4672",
        "IE_2009_4675",
        "IE_2009_4676",
        "IE_2009_4679",
        "IE_2009_4687A",
        "IE_2009_4690",
        "IE_2009_4702",
        "IE_2009_4703",
        "IE_2009_4707",
        "IE_2009_4713",
        "IE_2009_4726",
        "IE_2009_4731",
        "IE_2009_4734",
        "IE_2009_4753",
        "IE_2009_4754",
        "IE_2009_4763",
        "IE_2009_4767")

AS <- c("250",
        "251",
        "830",
        "881-11",
        "881-19",
        "903")

BAL <- c("94857-A",
         "94857-B",
         "N0078",
         "N0079",
         "N0080",
         "WAMZ444965",
         "WAMZ44962",
         "WAMZ44963")

BM <- c("E5492A",
        "E5492B",
        "E5492C",
        "E5492D",
        "E5492E",
        "E5524A",
        "E5524B",
        "E5524C",
        "E5524D",
        "E5524E")

BOU <- c("WAMZ44948",
         "WAMZ44949",
         "WAMZ44950",
         "WAMZ44951",
         "WAMZ44953",
         "WAMZ44954",
         "WAMZ44955",
         "WAMZ44956",
         "WAMZ44957",
         "WAMZ44958")

BS <- c("S6316A",
        "S6316B",
        "S6338A",
        "S6338B",
        "S6760-A",
        "S6760-B",
        "S6905A",
        "S6905B")

DB <- c("S4306",
        "S4307",
        "S4308",
        "S4309",
        "S4310")

DS <- c("PNG703",
        "PNG708",
        "PNG710")

EI <- c("S1312H",
        "S1312I",
        "S6741",
        "S6743",
        "S6775A")

EWS <- c("WAMZ88554",
         "WAMZ88558",
         "WAMZ88559",
         "WAMZ88570",
         "WAMZ88571",
         "WAMZ88576",
         "WAMZ88577",
         "WAMZ88579",
         "WAMZ88583",
         "WAMZ88584")

HB <- c("S4472K",
        "S4472L",
        "S4472M",
        "S4472N",
        "S4472O")

LAR <- c("WAMZ88555",
         "WAMZ88556",
         "WAMZ88563",
         "WAMZ88568",
         "WAMZ88572",
         "WAMZ88573",
         "WAMZ88574",
         "WAMZ88575",
         "WAMZ88578")

PB <- c("AAD107",
        "AAD141")

RS <- c("140217-A",
        "140217-B",
        "140217-C",
        "36968-A",
        "36968-B",
        "36968-C",
        "36968-E",
        "36968-F",
        "37157-A",
        "93825-B",
        "93825-C",
        "94825-A",
        "94858-A",
        "94866-A",
        "94866-B",
        "94869-A",
        "94869-B",
        "N0071",
        "N0072",
        "N0073",
        "N0074",
        "N0075",
        "PDIVA-C",
        "PDIVA-D",
        "PDIVA-E",
        "PDIVA-G",
        "PDIVA-I")

SCO <- c("N0121",
         "N0122",
         "N0123",
         "N0125")

SG <- c("E6408",
        "E6420")

SHE <-c("S5741-B",
        "S5741-C",
        "S5741-F",
        "S5742",
        "S5743")

SSI<- c("S0185",
        "S0460",
        "S0539-B",
        "S0539-D",
        "S0990A",
        "S0990C",
        "WAMZ44590",
        "WAMZ44932")

SWS <- c("WAMZ88585",
         "WAMZ88586",
         "WAMZ88587",
         "WAMZ88588",
         "WAMZ88590",
         "WAMZ88591",
         "WAMZ88592",
         "WAMZ88593",
         "WAMZ88594",
         "WAMZ88595",
         "WAMZ88596",
         "WAMZ88597",
         "WAMZ88598",
         "WAMZ88599")

GENOME.class <- set.populations(GENOME.class,list(RS,	AS,	BAL, PB,	BM,	SG,	AL,	SCO,	DS,	SSI, EI, DB,	HB,	SHE, BS, BOU,	EWS, LAR,	SWS), diploid=T)

GENOME.class@populations

#calculate neutrality tests
GENOME.class <- neutrality.stats(GENOME.class, detail=TRUE)

get.neutrality(GENOME.class)

TajD <- as.table(GENOME.class@Tajima.D)
write.csv(TajD, "./ophionotus/analysis/genetic_statistics/popgenome/vic/TajD.csv")

FuFs <- as.table(GENOME.class@Fu.F_S)
write.csv(FuFs, "./ophionotus/analysis/genetic_statistics/popgenome/vic/FuFs.csv")
```

Calculate Tajima’s D and Fu’s FS between localities of O. hexactis 

```{r, eval=F, echo=T}
library(PopGenome)
GENOME.class <- readData("./ophionotus/analysis/genetic_statistics/popgenome/hex_thin140.vcf", format="VCF", include.unknown=TRUE)
get.sum.data(GENOME.class)
GENOME.class@populations

get.individuals(GENOME.class)

hex_BM <- c("E5493A",
            "E5493B",
            "E5493C",
            "E5493D",
            "E5493E",
            "E5493F",
            "E5493G",
            "E5493H",
            "E5493I",
            "E5493J")

hex_SG <- c("S0036A",
            "S0036B",
            "S0036C",
            "S0036D",
            "S0036E",
            "S0036F",
            "S0036G",
            "S0036H",
            "S0036I",
            "S0036J")

hex_SR <- c("S3070",
            "S3533O",
            "S3753",
            "S3754",
            "S3755",
            "S3756",
            "S3757",
            "S3758",
            "S3759",
            "S3760")

hex_HI <- c("WAMZ43230",
            "WAMZ43231",
            "WAMZ43232",
            "WAMZ43233",
            "WAMZ43234",
            "WAMZ43235",
            "WAMZ43236",
            "WAMZ43237",
            "WAMZ43238",
            "WAMZ43239")

GENOME.class <- set.populations(GENOME.class,list(hex_BM, hex_SG, hex_SR, hex_HI), diploid=T)

GENOME.class@populations

#calculate neutrality tests
GENOME.class <- neutrality.stats(GENOME.class, detail=TRUE)

get.neutrality(GENOME.class)

TajD <- as.table(GENOME.class@Tajima.D)
write.csv(TajD, "./ophionotus/analysis/genetic_statistics/popgenome/hex/TajD.csv")

FuFs <- as.table(GENOME.class@Fu.F_S)
write.csv(FuFs, "./ophionotus/analysis/genetic_statistics/popgenome/hex/FuFs.csv")
```

## PCA within and between species 
```{r, eval=F, echo=T}
library(adegenet)
library(vcfR)
library(dartR)
setwd("./ophionotus/analysis/PCA")

##PCA with all samples from O. victoriae and O. hexactis

#vcf to genlight
vcf <- read.vcfR("./vic_hex_thin140.vcf.recode.vcf")
oph_genlight <- vcfR2genlight(vcf)
oph_genlight@pop

#export ind.names on genlight as species
ID_list <- as.data.frame(oph_genlight@ind.names)

write.csv(ID_list, file="Oph_195_ID_genlight.csv")

#Import popmap. Make sure the order the IDs in popmap is the same as the order in oph_genlight@ind.names 
pops_species <- read.csv('vic_hex_195_popmap.csv', header=TRUE)

strata(oph_genlight) <- pops_species
setPop(oph_genlight) <- ~species

oph_genlight@ind.names

#PCA on genlight objects

pca1 <- glPca(oph_genlight) #Number of axes selected: 100
pca1

print(pca1)
head(pca1)

#obtain variation per each component (PC1 and PC2 here)
#https://groups.google.com/forum/#!topic/poppr/yUdJycNYrhc

var_frac <- pca1$eig/sum(pca1$eig)
signif(sum(var_frac[1:1]) * 100, 4) #PC1

#[1] 2.688

signif(sum(var_frac[2:2]) * 100, 4) #PC2

#[1] 1.658

#Plot PCA

library(adegraphics)

s.class(pca1$scores, pop(oph_genlight),  col = c('#1f78b4', '#ff7f00'), paxes.draw = TRUE, ellipseSize = 0, plabels.cex = 0, key = list(space = "right"))
```

## DAPC between species to get information of alleles contributing the most to the discriminant functions
```{r, eval=F, echo=T}
library(adegenet)
library(vcfR)
library(dartR)
setwd("./ophionotus/analysis/DAPC")

vcf <- read.vcfR("./vic_hex_thin140.vcf.recode.vcf")
oph_genlight <- vcfR2genlight(vcf)
oph_genlight@pop

pops_species <- read.csv('./vic_hex_vicSG_hexBM_popmap.csv', header=TRUE) #Here we separate samples by victoriae, hexactis, victoriae from South Georgia, hexactis from Bransfield Mouth

strata(oph_genlight) <- pops_species
setPop(oph_genlight) <- ~species

#genlight to genind 

oph_genind <- gl2gi(oph_genlight, v = 1)

strata(oph_genind) <-data.frame(pops_species$species)

#We use find.clusters to identify clusters, although true clusters are, in this case, known (and accessible using pop(x)).
grp <- find.clusters(oph_genind, max.n.clust=50) #retain 200 PCs, 50 clusters 
names(grp)

table(pop(oph_genind), grp$grp)

#Rows correspond to actual groups (”ori”), while columns correspond to inferred groups (”inf”). 
#Here, we can see that original groups have nearly been perfectly identified by the method.

table.value(table(pop(oph_genind), grp$grp))

dapc1 <- dapc(oph_genind, grp$grp) #retain 200 PCs, 2 discriminant functions 
dapc1 #proportion of conserved variance = 1
set.seed(4)

pdf(file = 'loadingplot_contrib.pdf')

#set threshold of loading significance at 0.999 quantile
contrib <- loadingplot(dapc1$var.contr, axis=1, threshold=quantile(dapc1$var.contr,0.999), lab.jitter = 1)
contrib

#"CLocus-61907_100.A"  "CLocus-61907_100.T"  "CLocus-137845_119.A" "CLocus-137845_119.T"
#"CLocus-172167_99.A"  "CLocus-172167_99.T" 

dev.off()

freq_CLocus61907 <- tab(genind2genpop(oph_genind[loc=c("CLocus-61907_100")]), freq=TRUE)
freq_CLocus61907

freq_CLocus137845 <- tab(genind2genpop(oph_genind[loc=c("CLocus-137845_119")]), freq=TRUE)
freq_CLocus137845

freq_CLocus172167 <- tab(genind2genpop(oph_genind[loc=c("CLocus-172167_99")]), freq=TRUE)
freq_CLocus172167

freq_CLocus <- tab(genind2genpop(oph_genind[loc=c("CLocus-61907_100", "CLocus-137845_119", "CLocus-172167_99")]), freq=TRUE)
freq_CLocus


pdf(file = 'loading_plot_3loci.pdf')

par(mfrow=c(2,2), mar=c(5.1,4.1,4.1,.1),las=3)

matplot(freq_CLocus61907, pch=c("A", "T"), type="b",
        xlab="cluster",ylab="allele frequency", xaxt="n",
        cex=1.5, main="CLocus-61907")

axis(side=1, at=1:4, lab=c("O. victoriae (overall)","O. hexactis (BM)","O. victoriae (SG)","O. hexactis"))

matplot(freq_CLocus137845, pch=c("A", "T"), type="b",
        xlab="cluster",ylab="allele frequency", xaxt="n",
        cex=1.5, main="CLocus-137845")

axis(side=1, at=1:4, lab=c("O. victoriae (overall)","O. hexactis (BM)","O. victoriae (SG)","O. hexactis"))

matplot(freq_CLocus172167, pch=c("A", "T"), type="b",
        xlab="cluster",ylab="allele frequency", xaxt="n",
        cex=1.5, main="CLocus-172167")

axis(side=1, at=1:4, lab=c("O. victoriae (overall)","O. hexactis (BM)","O. victoriae (SG)","O. hexactis"))

dev.off()
```
## Population structure inference (STRUCTURE)

Run STRUCUTRE via structure_threader:
```{bash, error=TRUE, results='hide'}
setwd("./ophionotus/analysis/structure")

#Converted VCF to structure file via PGDSpider (from vic_hex_thin140.vcf.recode.vcf to vic_hex_stru.txt)

#Between O. victoriae and O. hexactis samples  (from vic_hex_thin140.vcf.recode.vcf to vic_hex_stru.txt)
./bin/structure_threader run -K 10 -R 10 -i ./vic_hex_stru.txt -o ./vic_hex_results -t 12 -st ./bin/structure

#Within O. victoriae samples (from vic_thin140.vcf.recode.vcf to vic_stru.txt)
./bin/structure_threader run -K 10 -R 10 -i ./vic_stru.txt -o ./vic_results -t 12 -st ./bin/structure

#Within O. hexactis samples  (from hex_thin140.vcf.recode.vcf to hex_stru.txt)
./bin/structure_threader run -K 10 -R 10 -i ./hex_stru.txt -o ./hex_results -t 12 -st ./bin/structure
```

Then export the STRUCTURE outputs to CLUMPAK (http://clumpak.tau.ac.il/) to generate consensus across replicates within each K

## Plotting STRUCTURE outputs 

Plot results using pophelper (https://github.com/royfrancis/pophelper)
```{r, eval=F, echo=T}
library(pophelper)
library(tcltk2)
library(gridExtra)
library(ggplot2)
library(label.switching)
library(tidyr)
library(gtable)

###Between O. victoriae and O. hexactis samples. Best K identified (Evanno method; k=2 , highest log likelihood; k=4)

#set path to the folder with CLUMPAK outputs
setwd("./ophionotus/analyses/structure/vic_hex/best_k_clumpak")

sfiles <- list.files(pattern ="./*.txt")

# include individual names from STRUCTURE file as row names
slist <- readQ(files=sfiles, indlabfromfile=T)

#Check whether everything is loaded properly
readQ(files=sfiles,filetype="structure")

# check class of ouput
class(slist)
head(slist[[1]])

# qlist attributes
attributes(slist)

# dataframe attributes
attributes(slist[[1]])

#tabulate Q
tr1 <- tabulateQ(qlist=slist)
tabulateQ(slist)

#SummariseQ
sr1 <- summariseQ(tr1)
summariseQ(tr1, writetable=TRUE)

#plot K=2 and K=4 results together
slist <- readQ(sfiles)

slist1 <- alignK(slist[c(1,2)])
p2 <- plotQ(slist1,imgoutput="join",returnplot=T,exportplot=F,basesize=11)
p2

#read pop file
vic_hex_poplabel <- read.csv("./vic_hex_popmap.csv", header=T,stringsAsFactors=F)

#plot with species labels

p4 <-plotQ(slist1,imgoutput="join", returnplot=T,exportplot=T,quiet=T, height=1.5, divsize=0.2,
           grplab = vic_hex_poplabel, subsetgrp=c("victoriae","hexactis"), selgrp="species", ordergrp=T, grplabangle = 90, grplabpos=1, grplabsize=0.8, grplabheight=0.1, indlabheight=0.1, indlabsize=2, showlegend=T, showindlab = F)


###Within O. victoriae samples. Best K identified (Evanno method; k=2 , highest log likelihood; k=3)

#set path to the folder with CLUMPAK outputs
setwd("./ophionotus/analyses//structure/vic/best_k_clumpak")

sfiles <- list.files(pattern ="./*.txt")

# include individual names from STRUCTURE file as row names
slist <- readQ(files=sfiles, indlabfromfile=T)

#Check whether everything is loaded properly
readQ(files=sfiles,filetype="structure")

# check class of ouput
class(slist)
head(slist[[1]])

# qlist attributes
attributes(slist)

# dataframe attributes
attributes(slist[[1]])

#tabulate Q
tr1 <- tabulateQ(qlist=slist)
tabulateQ(slist)

#SummariseQ
sr1 <- summariseQ(tr1)
summariseQ(tr1, writetable=TRUE)

#plot K=1 and K=2 results together
slist <- readQ(sfiles)

slist1 <- alignK(slist[c(1,2)])
p2 <- plotQ(slist1,imgoutput="join",returnplot=T,exportplot=F,basesize=11)
p2

#read pop file. Separate samples by locations
vic_poplabel <- read.csv("./vic_popmap.csv", header=T,stringsAsFactors=F)

#plot with location labels

colour_vic <- c("#00649B", "#52B3EB", "#E82F5F")

p5 <-plotQ(slist1,imgoutput="join", returnplot=T,exportplot=T,quiet=T, height=1.5, width=5.5, imgtype="pdf", dpi=600, sortind="all", sharedindlab=F, showdiv=T, divtype=1, divsize=0.2,
           grplab = vic_poplabel, subsetgrp=c("SG",
                                                "SSI",
                                                "HB",
                                                "DB",
                                                "EI",
                                                "SHE",
                                                "BS",
                                                "LAR",
                                                "SWS",
                                                "EWS",
                                                "AS",
                                                "RS",
                                                "SCO",
                                                "BAL",
                                                "AL",
                                                "TB",
                                                "PB",
                                                "BOU"), selgrp="location", clustercol=colour_vic, ordergrp=TRUE,
           grplabangle = 90, grplabpos=0.5, grplabsize=0.8, grplabheight=0.1, showlegend=T, showindlab = F)



###Within O. hexactis samples. Best K identified (Evanno method; k=2 , highest log likelihood; k=1)

#set path to the folder with CLUMPAK outputs
setwd("./ophionotus/analyses//structure/hex/best_k_clumpak")

sfiles <- list.files(pattern ="./*.txt")

# include individual names from STRUCTURE file as row names
slist <- readQ(files=sfiles, indlabfromfile=T)

#Check whether everything is loaded properly
readQ(files=sfiles,filetype="structure")

# check class of ouput
class(slist)
head(slist[[1]])

# qlist attributes
attributes(slist)

# dataframe attributes
attributes(slist[[1]])

#tabulate Q
tr1 <- tabulateQ(qlist=slist)
tabulateQ(slist)

#SummariseQ
sr1 <- summariseQ(tr1)
summariseQ(tr1, writetable=TRUE)

#plot K=1 and K=2 results together
slist <- readQ(sfiles)

slist1 <- alignK(slist[c(1,2)])
p2 <- plotQ(slist1,imgoutput="join",returnplot=T,exportplot=F,basesize=11)
p2

#read pop file. Separate samples by locations
hex_poplabel <- read.csv("./hex_popmap.csv", header=T,stringsAsFactors=F)

#plot with location labels
colour_hex <- c("#a6611a", "#dfc27d")

p4 <-plotQ(slist1,imgoutput="join", returnplot=T,exportplot=T,quiet=T, height=1.5, divsize=0.2, imgtype="pdf", dpi=600,
           grplab = hex_poplabel, subsetgrp=c("BM","SG","SR","HI"), selgrp="location", clustercol=colour_hex,
            ordergrp=T, grplabangle = 90, showdiv=FALSE, grplabpos=0.5, grplabsize=0.8, grplabheight=0.1, showlegend=T, showindlab = F)
```
