Outlier loci detection with OutFlank
================

## Setting up OutFLANK

See tutorial at
<https://adnguyen.github.io/2017_Ecological_Genomics/Tutorial/2017-04-03_Popgenomics5_outlier_annotations_DAPC_OutFLANK.html>

First modify functions so it can read input file properly
(<https://github.com/whitlock/OutFLANK/issues/19>)

``` r
library(OutFLANK)
library(vcfR)
library(adegenet)
library(LEA)

setwd("./ophionotus/analyse/outflank")

getFSTs_diploids = function(popNameList, SNPDataColumn){  
  #eliminating the missing data for this locus
  popnames=unlist(as.character(popNameList))
  popNameTemp=popnames[which(SNPDataColumn!=9)]
  snpDataTemp=SNPDataColumn[SNPDataColumn!=9]
  
  HetCounts <- tapply(snpDataTemp, list(popNameTemp,snpDataTemp), length)
  HetCounts[is.na(HetCounts)] = 0
  
  #Case: all individuals are genetically identical at this locus
  if(dim(HetCounts)[2]==1){
    return (list(He=NA,FST=NA, T1=NA, T2=NA,FSTNoCorr=NA, T1NoCorr=NA, T2NoCorr=NA,meanAlleleFreq = NA))
  }
  
  if(dim(HetCounts)[2]==2){
    if(paste(colnames(HetCounts),collapse="")=="01"){HetCounts=cbind(HetCounts,"2"=0)}
    if(paste(colnames(HetCounts),collapse="")=="12"){HetCounts=cbind("0"=0,HetCounts)} 
    if(paste(colnames(HetCounts),collapse="")=="02"){HetCounts=cbind(HetCounts[,1],"1"=0, HetCounts[,2])}
  }
  
  out = WC_FST_Diploids_2Alleles(HetCounts) 
  return(out)
}

MakeDiploidFSTMat<-function(SNPmat,locusNames,popNames){
  locusname <- unlist(locusNames)
  popname <- unlist(popNames)
  snplevs <- levels(as.factor(unlist(SNPmat)))
  if(any(!(snplevs%in%c(0,1,2,9)))==TRUE) {
    print("Error: Your snp matrix has a character other than 0,1,2 or 9")
    break
  }
  if (dim(SNPmat)[1] != length(popname)) {
    print("Error: your population names do not match your SNP matrix")
    break
  }
  if (dim(SNPmat)[2] != length(locusname)) {
    print("Error:  your locus names do not match your SNP matrix")
    break
  }
  writeLines("Calculating FSTs, may take a few minutes...")
  nloci <- length(locusname)
  FSTmat <- matrix(NA, nrow = nloci, ncol = 8)
  for (i in 1:nloci) {
    FSTmat[i, ] = unlist(getFSTs_diploids(popname, SNPmat[,i]))
    if (i%%10000 == 0) {
      print(paste(i, "done of", nloci))
    }
  }
  outTemp = as.data.frame(FSTmat)
  outTemp = cbind(locusname, outTemp)
  colnames(outTemp) = c("LocusName", "He", "FST", "T1", "T2", 
                        "FSTNoCorr", "T1NoCorr", "T2NoCorr", "meanAlleleFreq")
  return(outTemp)
}
```

## OutFLANK between O. victoirae VS O. hexactis including O. victoriae from South Georgia

``` r
vic_hex_vcf <- read.vcfR("./vic_hex_thin140.vcf.recode.vcf", verbose=FALSE)

#convert vcf to .geno format 
vic_hex_geno = vcf2geno("./vic_hex_thin140.vcf.recode.vcf")
vic_hex.geno_in <- read.fwf(vic_hex_geno, width=rep(1,195))
vic_hex.geno <- t(vic_hex.geno_in)

vic_hex_meta <- read.csv("vic_hex_popmap.csv", header=T) # Read in popmap, with O. victoriae from South Georgia listed as O. hexactis

vic_hex_CLocus <- read.table("vic_hex_loci.txt", header=F) # Read CLocus ID (from --012 of vcftools / *.012.pos)

#Calculate FST on the data
my_fst <- MakeDiploidFSTMat(vic_hex.geno, locusNames = vic_hex_CLocus$V1, popNames = vic_hex_meta$species)
head(my_fst)

plot(my_fst$FST, my_fst$FSTNoCorr, 
     xlim=c(-0.01,0.3), ylim=c(-0.01,0.3),
     pch=20)
abline(0,1) # no loci deviate from the linear relationship 

plot(my_fst$He, my_fst$FSTNoCorr, pch=20, col="grey")

hist(my_fst$FSTNoCorr, breaks=seq(0,0.5, by=0.01))

hist(my_fst$FSTNoCorr[my_fst$He>0.05], breaks=seq(0,0.6, by=0.001))


#Running Outflank 
out1 <- OutFLANK(FstDataFrame=my_fst, NumberOfSamples=2,
                 RightTrimFraction = 0.05, LeftTrimFraction = 0.05,
                 qthreshold = 0.01, Hmin = 0.1)

OutFLANKResultsPlotter(out1, withOutliers = TRUE,
                       NoCorr = TRUE, Hmin = 0.1, binwidth = 0.001, Zoom =
                         FALSE, RightZoomFraction = 0.2, titletext = NULL)
hist(out1$results$pvaluesRightTail)


outliers <- which(out1$results$OutlierFlag=="TRUE")
print(outliers) #28 outliers. Note that these numbers just refer to the rank order of the SNPs in your data matrix. 

#To figure out which genes (transcript IDs) these SNPs actually belong to, read vcf file
vcfann <- as.data.frame(getFIX(vic_hex_vcf))
outlierlist <- vcfann[outliers,]

write.csv(outlierlist, file="outlierlist_vic_hex.csv")
```

## OutFLANK within O. victoirae

``` r
vic_vcf <- read.vcfR("./vic_thin140.vcf.recode.vcf", verbose=FALSE)

#convert vcf to .geno format 
vic_geno = vcf2geno("./vic_thin140.vcf.recode.vcf")
vic.geno_in <- read.fwf(vic_geno, width=rep(1,158))
vic.geno <- t(vic.geno_in)

vic_meta <- read.csv("vic_popmap.csv", header=T) # Read in popmap, with O. victoriae separated from enviroments (deep, shelf, island)

vic_CLocus <- read.table("vic_loci.txt", header=F) # Read CLocus ID (from --012 of vcftools / *.012.pos)

#Calculate FST on the data
my_fst <- MakeDiploidFSTMat(vic.geno, locusNames = vic_CLocus$V1, popNames = vic_meta$ecotype)
head(my_fst)

plot(my_fst$FST, my_fst$FSTNoCorr, 
     xlim=c(-0.01,0.3), ylim=c(-0.01,0.3),
     pch=20)
abline(0,1) # no loci deviate from the linear relationship 

plot(my_fst$He, my_fst$FSTNoCorr, pch=20, col="grey")

hist(my_fst$FSTNoCorr, breaks=seq(0,0.5, by=0.01))

hist(my_fst$FSTNoCorr[my_fst$He>0.05], breaks=seq(0,0.6, by=0.001))

#Running Outflank 
out2 <- OutFLANK(FstDataFrame=my_fst, NumberOfSamples=3,
                 RightTrimFraction = 0.03, LeftTrimFraction = 0.05,
                 qthreshold = 0.01, Hmin = 0.1)

OutFLANKResultsPlotter(out2, withOutliers = TRUE,
                       NoCorr = TRUE, Hmin = 0.1, binwidth = 0.005, Zoom =
                         FALSE, RightZoomFraction = 0.1, titletext = NULL)

hist(out2$results$pvaluesRightTail)


outliers <- which(out2$results$OutlierFlag=="TRUE")
print(outliers) #3 outliers

vcfann <- as.data.frame(getFIX(vic_vcf))
outlierlist <- vcfann[outliers,]

write.csv(outlierlist, file="outlierlist_vic.csv")
```

## OutFLANK within O. hexactis

``` r
hex_vcf <- read.vcfR("./hex_thin140.vcf.recode.vcf", verbose=FALSE)

#convert vcf to .geno format 
hex_geno = vcf2geno("./hex_thin140.vcf.recode.vcf")
hex.geno_in <- read.fwf(hex_geno, width=rep(1,40))
hex.geno <- t(hex.genoo_in)

hex_meta <- read.csv("hex_popmap.csv", header=T) # Read in popmap, with O. hexactis separated by locations

hex_CLocus <- read.table("hex_loci.txt", header=F) # Read CLocus ID (from --012 of vcftools / *.012.pos)

#Calculate FST on the data
my_fst <- MakeDiploidFSTMat(hex.geno, locusNames = hex_CLocus$V1, popNames = hex_meta$location)
head(my_fst)

plot(my_fst$FST, my_fst$FSTNoCorr, 
     xlim=c(-0.01,0.3), ylim=c(-0.01,0.3),
     pch=20)
abline(0,1) # no loci deviate from the linear relationship 

plot(my_fst$He, my_fst$FSTNoCorr, pch=20, col="grey")

hist(my_fst$FSTNoCorr, breaks=seq(0,0.5, by=0.01))

hist(my_fst$FSTNoCorr[my_fst$He>0.05], breaks=seq(0,0.6, by=0.001))

#Running Outflank 
out3 <- OutFLANK(FstDataFrame=my_fst, NumberOfSamples=4,
                 RightTrimFraction = 0.05, LeftTrimFraction = 0.05,
                 qthreshold = 0.01, Hmin = 0.1)

OutFLANKResultsPlotter(out3, withOutliers = TRUE,
                       NoCorr = TRUE, Hmin = 0.1, binwidth = 0.005, Zoom =
                         FALSE, RightZoomFraction = 0.05, titletext = NULL)

hist(out3$results$pvaluesRightTail)


outliers <- which(out3$results$OutlierFlag=="TRUE")
print(outliers) #0 outlier
```
